threaddb/mail.go

```diff
@@ -0,0 +1,74 @@
+package threaddb
+
+import (
+	"context"
+
+	"github.com/alecthomas/jsonschema"
+	dbc "github.com/textileio/go-threads/api/client"
+	"github.com/textileio/go-threads/core/thread"
+	db "github.com/textileio/go-threads/db"
+	"github.com/textileio/textile/api/common"
+	"github.com/textileio/textile/mail"
+)
+
+var (
+	mailSchema  *jsonschema.Schema
+	mailIndexes = []db.Index{{
+		Path: "from",
+	}, {
+		Path: "to",
+	}, {
+		Path: "created_at",
+	}, {
+		Path: "read_at",
+	}}
+	mailConfig db.CollectionConfig
+)
+
+// Message represents the mail threaddb collection schema.
+type Message struct {
+	ID        string `json:"_id"`
+	From      string `json:"from"`
+	To        string `json:"to"`
+	Body      string `json:"body"`
+	Signature string `json:"signature"`
+	CreatedAt int64  `json:"created_at"`
+	ReadAt    int64  `json:"read_at"`
+}
+
+func init() {
+	reflector := jsonschema.Reflector{ExpandedStruct: true}
+	mailSchema = reflector.Reflect(&Message{})
+	mailConfig = db.CollectionConfig{
+		Name:    mail.CollectionName,
+		Schema:  mailSchema,
+		Indexes: mailIndexes,
+	}
+}
+
+// Mail is a wrapper around a threaddb collection for sending mail between users.
+type Mail struct {
+	collection
+}
+
+// NewMail returns a new mail collection mananger.
+func NewMail(tc *dbc.Client) (*Mail, error) {
+	return &Mail{
+		collection: collection{
+			c:      tc,
+			config: mailConfig,
+		},
+	}, nil
+}
+
+// NewMailbox creates a new threaddb mail box.
+func (m *Mail) NewMailbox(ctx context.Context, name string, opts ...Option) (thread.ID, error) {
+	args := &Options{}
+	for _, opt := range opts {
+		opt(args)
+	}
+	id := thread.NewIDV1(thread.Raw, 32)
+	ctx = common.NewThreadNameContext(ctx, name)
+	err := m.c.NewDB(ctx, id, db.WithNewManagedName(name), db.WithNewManagedCollections(mailConfig), db.WithNewManagedToken(args.Token))
```

Comment:
Q since my memory on this is old:
This `NewMailbox` was used to have idempotency behavior. A priori before the call a `NewIDV1` is generated.
Why `NewDB` call would fail if called more than once? Can't you always have an identical collection in a new thread?

Reply:
Oh, this is more interceptor magic... making this call here results in a thread being tracked (along with the name) for the owner. Subsequent calls will error with a unique constraint violation. The unique constraint on `mongodb.Thread` is fired before the DB is created in `go-threads`.

Reply:
Ah, so uniqueness is done by Hub, not a threadsd.

Reply:
Yep, uniqueness by name is done by the Hub